modules/internal/ChapelLocale.chpl

```diff
@@ -90,20 +90,142 @@ module ChapelLocale {
     return (subloc != c_sublocid_none
             && subloc != c_sublocid_any
             && subloc != c_sublocid_all);
+  
+  /*
+    regular: Has a concrete BaseLocale instance
+    nilLocale: The _instance is set to nil. Used during setup. Also, as a
+               sentinel value in locale tree operations
+    dummy: Used during initialization for `here` before it is properly setup
+    default: Used to store the default locale instance. Initially set to nil,
+             then "fixed" by LocalesArray to Locales[0]
+    any: Placeholder to represent the notion of "anywhere"
+   */
+  pragma "no doc"
+  enum localeKind { regular, any, nilLocale, dummy, default };
+
+  const nilLocale = new locale(localeKind.nilLocale);
+  var defaultLocale = new locale(localeKind.default);
+
+  pragma "always RVF"
+  record _locale {
+    var _instance: unmanaged BaseLocale?;
+
+    inline proc _value {
+      return _instance!;
+    }
+
+    forwarding _value;
+
+    // default initializer for the locale record.
+    // TODO: What is the default value for a locale?
+    proc init() {
+      if doneCreatingLocales {
+        this._instance = defaultLocale._instance;
+      }
+      else {
+        this._instance = new unmanaged LocaleModel();
+      }
+    }
+
+    // used internally during setup
+    pragma "no doc"
+    proc init(_instance: BaseLocale) {
+      this._instance = _to_unmanaged(_instance);
+    }
+
+    pragma "no doc"
+    proc init(param kind) {
+      if kind == localeKind.regular then
+        compilerError("locale.init(kind) can not be used to create ",
+                      "a regular locale instance");
+      else if kind == localeKind.dummy then
+        this._instance = new unmanaged DummyLocale();
+      else if kind == localeKind.default then
+        this._instance = nil;
+    }
+
+    proc init=(other: locale) {
+      this._instance = other._instance;
+    }
+
+    proc deinit() { }
+
+    inline proc maxTaskPar { return this._instance!.maxTaskPar; }
+    inline proc callStackSize { return this._instance!.callStackSize; }
+
+    // the following are normally taken care of by `forwarding`. However, they
+    // don't work if they are called in a promoted expression. See 15148
+    inline proc numPUs(logical: bool = false, accessible: bool = true) {
+      return this._instance!.numPUs();
```

Comment:
Need to pass the incoming arguments through: `this._instance!.numPUs(logical, accessible)`.